### ISSUE TITLE
M-backspace does not effectively delete words in ansi-term

### ISSUE THREAD

**Issue by wizmer** [pmarinov/clean-aindent-mode#8]
#### Description :octocat:
M-backspace visually delete a word in ansi-term but when Enter is pressed the deleted word is still part of the command.

#### Reproduction guide :beetle:
- Start Emacs
- M-x ansi-term
- Type "ls"
- Delete it using M-backspace
- press Enter

*Observed behaviour:* :eyes: :broken_heart:
The "ls" command is performed even it had been deleted

*Expected behaviour:* :heart: :smile:
The "ls" command should not be performed.
The problem is that M-backspace is overriden by clean-aindent-mode and set to (clean-aindent--bsunindent ARG)

Maybe clean-aindent should always be disabled when being in term-mode ?

PS: The issue was originally opened here: syl20bnr/spacemacs#9564

**Comment by pmarinov** [comment]
Thank you for submitting your fix.

Isn't it better if the entire function `clean-aindent--bsunindent(arg)` checks for `term-mode` and bypasses its present functionality?

I think your verification snippet needs to be immediately after `(interactive "p")`, because the entire functionality of Backspace Unindent has no meaning in `term-mode`.

I propose you change it as:

```
(interactive "p")
  (if (eq 'term-mode major-mode)
      ;; Special case, bypass functionality on 'term-mode
      (term-send-raw-meta)
    ;; else: all other editing modes are OK
    (if (not (clean-aindent--inside-indentp))
    etc...
```

What do you think?

**Comment by wizmer** [comment]
I don't know exactly what the whole mode is doing (I discovered it while trying to understand why M-backspace was not working), but I think it would even make sense to disable the whole mode in `term-mode`
I don't know if this is doable why keeping the mode as `:global`

**Comment by pmarinov** [comment]
Yes, you are right, disabling the entire mode inside "term-mode" is best.

I added this to my "init.el" to test the idea:
```
(defun term-mode-customize()
  ;; Clean AIndent Mode conflicts with functionality of "term-mode"
  (clean-aindent-mode -1))
(add-hook 'term-mode-hook 'term-mode-customize)
```
Then starting new `M-x term` has Alt-Backspace work as originally intended without the conflict.

Can this be added as part of the setup of _spacemacs_?

I can also add this snippet to the documentation (README) of `clean-aindent-mode`.

**Comment by wizmer** [comment]
Yep, I am currently using this hook as well.
But the mode is supposed to behave correctly without the user intervention.
I just discovered the derived modes: https://www.gnu.org/software/emacs/manual/html_node/elisp/Basic-Major-Modes.html#Basic-Major-Modes
We could could activate clean-aindent-mode only for modes that derive from prog-modes ?
What do you think ?

**Comment by aaronjensen** [comment]
Hm, can you explain why that works @wizmer? I came here to file the same bug. I fixed it as @pmarinov suggested, but, because it's a global minor mode this disables it entirely whenever you first enter a term, so that's not ideal.

**Comment by aaronjensen** [comment]
I think I understand it now, remap actually causes the command to take over whenever `backward-kill-word` would have been called. That's super helpful to know and does seem like the right solution to this.

**Comment by pmarinov** [comment]
I like this idea:
`(define-key clean-aindent-mode--keymap [remap backward-kill-word] 'clean-aindent--bsunindent)`

I will test that myself now.

**Comment by pmarinov** [comment]
It works. I tested with `emacs -q` and manually `(load-file "new-clean-aindent-mode.el")` and activated the *mode* to make sure in a clean environment that the idea is good.

Now, @wizmer if you submit a pull request with this change **I will merge it**. Thank you for this very nice solution.

**Comment by aaronjensen** [comment]
@pmarinov https://github.com/pmarinov/clean-aindent-mode/pull/9

**Comment by pmarinov** [comment]
It has been merged. _Melpa_ will pick it up automatically, let's see how it goes.
:-)
